src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,107 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int STARTING_CAPACITY = 10;
+    private static final double INCREASE_FACTOR = 1.5;
+
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[STARTING_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkCapacity();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        checkCapacity();
+
+        System.arraycopy(elements, index, elements, index + 1, size - index);
 
+        elements[index] = value;
+        size++;
     }
 
-    @Override
     public void addAll(List<T> list) {
-
+        if (list != null) {
+            for (int i = 0; i < list.size(); i++) {
+                add(list.get(i));
+            }
+        }
```

Comment:
we want to get NPE if list is actually null rather silently skip this method
```suggestion
        for (int i = 0; i < list.size(); i++) {
            add(list.get(i));
        }
```